dbms/src/Core/Block.h

```diff
@@ -35,6 +35,9 @@ class Block
 
 public:
     BlockInfo info;
+    /// Input stream could use delayed_defaults to add addition info at which rows it have inserted default values.
+    /// Such values would be replaced later by column defaults in AddingDefaultsBlockInputStream (if any).
+    BlockDelayedDefaults delayed_defaults;
```

Comment:
I think that it's better to avoid adding delayed_defaults inside block because they are used only on client for special purpose. But it may be difficult to change interface of `BlockInputStream`, so may be it's ok.

As for me, it would be better to calculate defaults in `BlockInputStreamFromRowInputStream`. Then you won't need `AddingDefaultBlockInputStream`, and won't need to add delayed defaults into block.

Reply:
As I could understand we can't add defaults in BlockInputStreamFromRowInputStream cause of InputStreamFromASTInsertQuery and other remote cases. There could be other input streams there.